### PR TITLE
Add checks for profile picture while uploading #121

### PR DESCRIPTION
**Pull request by somya51p**
# Related Issue
Fixes: 

# Prosposed Changes
- change 1
- change 2

# Additional Info
- any additional information or context

# Screenshots

Original             |  Updated
:-------------------------:|:-------------------------:
** original screenshot **  |  ** updated screenshot **

### CODE REVIEW

**Comment by garg3133** [comment]
Good work @somya51p, but we don't need our backend to decide whether the uploaded file can be accepted as a profile image or not, all this work must be done in frontend only. One more issue with going in backend is that when we come back to the frontend with the error, all the data entered by the user will be gone because of one error and the user will need to fill the data again which is really bad for UX.

Also, you should use appropriate variable names for everything (you've used `d` at some places and use of `error` is also wrong as it doesn't tell us which type of error is it).

You can create a function in the template's script and call that function when the file input's `onChange` event is fired. And if a wrong file is selected, show an alert and remove that file from the file input.

File: accounts/views.py [hunk 1]
```diff
@@ -11,7 +11,6 @@
 from django.utils.encoding import force_bytes, force_text
 from django.utils.html import strip_tags
 from django.utils.http import urlsafe_base64_encode, urlsafe_base64_decode
-
```
Comment (garg3133) [inline accounts/views.py:14]: Revert these to their previous state (you've deleted some newlines in these files).

File: templates/accounts/volunteer_profile.html [hunk 2]
```diff
@@ -147,7 +147,7 @@ <h2 class="font-weight-bolder">Complete Your Profile</h2>
                 <!-- Add profile picture -->
                 <div class="mb-3">
                     <label for="profileImage">Profile Picture <span class="text-muted">(Optional)</span></label>
-                    <input name="profile_image" type="file" id="profileImage" class="form-control-file">
+                    <input name="profile_image" type="file" id="profileImage" class="form-control-file" onchange="return ImageValidation()">
```
Comment (garg3133) [inline templates/accounts/volunteer_profile.html:150]: No need to add `return` here. You just want to execute the function when onChange event is fired and not return anything. Return is important when this function is called on `onsubmit` event of form element.

File: templates/accounts/volunteer_profile.html [hunk 3]
```diff
@@ -229,7 +229,18 @@ <h2 class="font-weight-bolder">Complete Your Profile</h2>
         }, false)
     }())
 </script>
-
+<script type="text/javascript">
+    function ImageValidation(){
```
Comment (garg3133) [inline templates/accounts/volunteer_profile.html:233]: function names shouldn't start with a capital letter. Only class names start with capital letters.

File: templates/accounts/volunteer_profile.html [hunk 4]
```diff
@@ -229,7 +229,18 @@ <h2 class="font-weight-bolder">Complete Your Profile</h2>
...
+    var fileInput = document.getElementById('profileImage');
+    var filePath = fileInput.value;
+    var allowedExtensions = /(\.jpg|\.jpeg|\.png|\.gif)$/i;
+    if(!allowedExtensions.exec(filePath)){
+        alert('Please upload file having extensions .jpeg, .jpg, .png or .gif only.');
+        fileInput.value = '';
+        return false;
+    }
+}
```
Comment (garg3133) [inline templates/accounts/volunteer_profile.html:242]: Follow proper indentations here.

File: templates/accounts/volunteer_profile.html [hunk 5]
```diff
@@ -229,7 +229,18 @@ <h2 class="font-weight-bolder">Complete Your Profile</h2>
...
-
+<script type="text/javascript">
+    function ImageValidation(){
+    var fileInput = document.getElementById('profileImage');
+    var filePath = fileInput.value;
+    var allowedExtensions = /(\.jpg|\.jpeg|\.png|\.gif)$/i;
+    if(!allowedExtensions.exec(filePath)){
+        alert('Please upload file having extensions .jpeg, .jpg, .png or .gif only.');
+        fileInput.value = '';
+        return false;
```
Comment (garg3133) [inline templates/accounts/volunteer_profile.html:240]: No need to return anything here. This is important only when it is used in `onsubmit` event (to stop the form from submitting).

**Review by garg3133 — COMMENTED** [review]
This is exactly what I wanted, so good work with that. I've posted a bunch of reviews, so please see those.

After you are done with the changes, please squash all your commits into a single commit. This helps us keep the commit history clean and is very useful while examining the source of a bug and what caused it in the commit history. Follow these instructions to do so:
* `git rebase -i HEAD~<no. of commits you made in total>`
* Except for first `pick`, replace all `pick` with `squash` in the file opened.
* Close the file after saving it.
* In the new file opened, remove all commit messages and write a new one and then close this file too after saving.
* Force push your changes. (`git push origin +<your-branch-name>`).

**Comment by garg3133** [comment]
One more thing, you should never ever make any changes directly in the master branch. Always create a new branch before making any changes and open a pull request from that branch only. The master branch should always be in sync with the master branch of the main repo.

**Comment by somya51p** [comment]
@garg3133 sir I have made the changes and squashed the commits into one. Next time I will keep in mind to make changes making new branch. Thanks for the guidance. Review the same.

File: accounts/views.py [hunk 6]
```diff
@@ -12,6 +12,7 @@
 from django.utils.html import strip_tags
 from django.utils.http import urlsafe_base64_encode, urlsafe_base64_decode
 
+
```
Comment (garg3133) [inline accounts/views.py:15]: Remove this extra line. There should be a single newline only between the two sections of import. Also, this `accounts/views.py` file shouldn't be a part of this PR as no changes are being done in this file.

File: accounts/views.py [hunk 7]
```diff
@@ -156,7 +157,7 @@ def complete_profile(request):
         dob = request.POST['dob']
         contact_no = request.POST['contact_no']
         profile_image = request.FILES.get('profile_image')
-
+
```
Comment (garg3133) [inline accounts/views.py:160]: Remove these trailing whitespaces you added (keeping the newline).

File: templates/accounts/volunteer_profile.html [hunk 8]
```diff
@@ -229,7 +229,17 @@ <h2 class="font-weight-bolder">Complete Your Profile</h2>
...
+    function imageValidation() {
+        var fileInput = document.getElementById('profileImage');
+        var filePath = fileInput.value;
+        var allowedExtensions = /(\.jpg|\.jpeg|\.png|\.gif)$/i;
+        if(!allowedExtensions.exec(filePath)) {
+            alert('Please upload file having extensions .jpeg, .jpg, .png or .gif only.');
+            fileInput.value = '';
+        }
+    }
+</script>
```
Comment (garg3133) [inline templates/accounts/volunteer_profile.html:242]: Add a single newline before and after this `script` block.

File: templates/volunteers/update_profile.html [hunk 9]
```diff
@@ -172,7 +173,17 @@ <h3 class="text-center">Update Profile</h3>
...
+    function imageValidation() {
+        var fileInput = document.getElementById('profileImage');
+        var filePath = fileInput.value;
+        var allowedExtensions = /(\.jpg|\.jpeg|\.png|\.gif)$/i;
+        if(!allowedExtensions.exec(filePath)) {
+            alert('Please upload file having extensions .jpeg, .jpg, .png or .gif only.');
+            fileInput.value = '';
+        }
+    }
+</script>
```
Comment (garg3133) [inline templates/volunteers/update_profile.html:186]: Add a single newline before and after this `script` block.

File: templates/volunteers/update_profile.html [hunk 10]
```diff
@@ -183,4 +194,4 @@ <h3 class="text-center">Update Profile</h3>
     $('#dob').datepicker("setDate", "{{ volun.dob|date:'Y-m-d' }}");
 </script>
 
-{% endblock %}
\ No newline at end of file
+{% endblock %}
```
Comment (garg3133) [inline templates/volunteers/update_profile.html:197]: No trailing whitespace should be there. There should just be a single newline at the end of the file.

**Review by garg3133 — COMMENTED** [review]
You haven't squashed the commits. There are still 4 commits in this PR.

One more thing, can you do the same changes on `students/new` page as well?

**Comment by somya51p** [comment]
@garg3133 sir, Please review the same. I have made the needful changes.

**Comment by somya51p** [comment]
@garg3133 sir, Have a look at the changes.

**Comment by garg3133** [comment]
Thanks a lot, @somya51p for working on this! Merging it into master 🎉

**Pull request merged**